In [1]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
d1 = 16  # last digit
d2 = 10  # second last digit

A = np.array([[d1 + 2, d2 + 1],
              [2*d1, d2 + 2]])

print("Matrix A:\n", A)



Matrix A:
 [[18 11]
 [32 12]]


In [4]:
# Shape
shape_A = A.shape

# Determinant
det_A = np.linalg.det(A)

# Rank
rank_A = np.linalg.matrix_rank(A)

# Eigenvalues
eigenvalues, eigenvectors = np.linalg.eig(A)

# Inverse (only if determinant != 0)
if det_A != 0:
    inv_A = np.linalg.inv(A)
else:
    inv_A = "Matrix is singular, cannot compute inverse"

# Display results
print("Shape of A:", shape_A)
print("Determinant of A:", det_A)
print("Rank of A:", rank_A)
print("Eigenvalues of A:", eigenvalues)
print("Eigenvectors of A:\n", eigenvectors)
print("Inverse of A:\n", inv_A)

Shape of A: (2, 2)
Determinant of A: -135.99999999999991
Rank of A: 2
Eigenvalues of A: [34. -4.]
Eigenvectors of A:
 [[ 0.56652882 -0.4472136 ]
 [ 0.82404192  0.89442719]]
Inverse of A:
 [[-0.08823529  0.08088235]
 [ 0.23529412 -0.13235294]]


In [5]:
B = A.copy()
B[0, 0] = B[0, 0] + 1

print("Matrix B:\n", B)

Matrix B:
 [[19 11]
 [32 12]]


In [7]:
# Shape
shape_B = B.shape

# Determinant
det_B = np.linalg.det(B)

# Rank
rank_B = np.linalg.matrix_rank(B)

# Eigenvalues
eigenvalues, eigenvectors = np.linalg.eig(B)

# Inverse (only if determinant != 0)
if det_B != 0:
    inv_B = np.linalg.inv(B)
else:
    inv_B = "Matrix is singular, cannot compute inverse"

# Display results
print("Shape of B:", shape_B)
print("Determinant of B:", det_B)
print("Rank of B:", rank_B)
print("Eigenvalues of B:", eigenvalues)
print("Eigenvectors of B:\n", eigenvectors)
print("Inverse of B:\n", inv_B)

Shape of B: (2, 2)
Determinant of B: -124.00000000000003
Rank of B: 2
Eigenvalues of B: [34.58533468 -3.58533468]
Eigenvectors of B:
 [[ 0.57663377 -0.43786931]
 [ 0.81700275  0.89903863]]
Inverse of B:
 [[-0.09677419  0.08870968]
 [ 0.25806452 -0.15322581]]


## 1. How did the determinant change and why?
The determinant changed from −136 to −124.
This happened because increasing one element alters the area-scaling factor of the matrix. Since determinants depend linearly on matrix entries, even a +1 change shifts the determinant.

## 2. Did the rank change?
No.Both matrices have rank 2. The rows and columns remain linearly independent, so the rank is unchanged.

## 3. How did the eigenvalues respond to the value change?
The eigenvalues changed slightly:
The larger eigenvalue increased and the smaller which is negative eigenvalue moved closer to zero.This is expected because eigenvalues are sensitive to matrix entry changes, but small value changes usually cause small eigenvalue shifts, not drastic jumps.

## 4. Is B easier or harder to invert than A? Why?
B is slightly easier to invert than A because the absolute determinant of B (124) is smaller than A (136), but both are far from zero. Since neither matrix is close to being singular, inversion is stable in both cases.